### ISSUE TITLE
MomentSGD has no attribute prepare

### ISSUE THREAD

**Issue by mengliu1991** [chainer/chainer#5687]
Recently, I run the code released by other authors. They utilized chainer v1.3, but I installed v4. When I run the code, it errors that Attribute Errors: MomentSGD optimizer has no attribute prepare. Here I post the codes of this part:

    class BaseModel(chainer.Chain):
      loss     = None
      accuracy = None
      gpu_mode = False
      _train   = False
      def __call__(self, *arg_list, **arg_dict):
        raise NotImplementedError()

      def clear(self):
        self.loss = None
        self.accuracy = None

      def train(self, data, optimizer):
        self._train = True
        optimizer.update(self, data)
        if self.accuracy is None:
            return float(self.loss.data)
        else:
            return float(self.loss.data), float(self.accuracy.data)

      def validate(self, data):
        self._train = False
        self(data)
        if self.accuracy is None:
            return float(self.loss.data)
        else:
            return float(self.loss.data), float(self.accuracy.data)

      def test(self, data):
        self._train = False
        raise NotImplementedError()

      def save(self, fname):
        serializers.save_hdf5(fname, self)

      def load(self, fname):
        serializers.load_hdf5(fname, self)

      def cache(self):
        self.to_cpu()
        cached_model = self.copy()
        self.to_gpu()
        return cached_model
      # this part is the error part
      def setup(self, optimizer):
        self.to_gpu()
        optimizer.target = self
        optimizer.prepare()

      def to_cpu(self):
        if not self.gpu_mode:
            return
        super(BaseModel, self).to_cpu()
        self.gpu_mode = False

      def to_gpu(self):
        if self.gpu_mode:
            return
        super(BaseModel, self).to_gpu()
        self.gpu_mode = True

**Comment by fiarabbit** [comment]
According to the [contribution guide](http://docs.chainer.org/en/stable/contribution.html), any general question is preferred to be raised at the StackOverflow with a `chainer` tag.

```
Warning

If you have a question on usages of Chainer, it is highly recommended to send a post to StackOverflow or Chainer User Group instead of the issue tracker. The issue tracker is not a place to share knowledge on practices. We may suggest these places and immediately close how-to question issues.
```

**Comment by niboshi** [comment]
Closing as the question has been posted on StackOverflow.
https://stackoverflow.com/questions/53404722/atrributeerror-momentsgd-optimizer-has-no-attribute-prepare
